compression_verifier.go

```diff
@@ -0,0 +1,224 @@
+			algorithm = strings.ToUpper(algorithm)
+			tableCompressions[table][column] = algorithm
+
+			if _, ok := c.supportedAlgorithms[algorithm]; !ok {
+				return &UnsupportedCompressionError{
+					table:     table,
+					column:    column,
+					algorithm: algorithm,
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// GetHashes compares the source data with the target data to ensure the integrity of the
+// data being copied.
+//
+// The CompressionVerifier method checks if the existing table contains compressed data
+// and will apply the decompression algorithm to the applicable columns if necessary.
+// After the columns are decompressed, the hashes of the data are used and verified
+// for comparison equality
+func (c *CompressionVerifier) GetHashes(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64, tableCompression map[string]string) (map[uint64][]byte, error) {
+	// Extract the raw rows using SQL to be decompressed
+	rows, err := c.getRows(db, schema, table, pkColumn, columns, pks)
+	if err != nil {
+		return nil, err
+	}
+
+	// Decompress applicable columns and hash the resulting column values for comparison
+	return c.decompressAndHash(table, columns, rows, tableCompression)
+}
+
+func (c *CompressionVerifier) getRows(db *sql.DB, schema, table, pkColumn string, columns []schema.TableColumn, pks []uint64) (*sql.Rows, error) {
+	quotedPK := quoteField(pkColumn)
+	sql, args, err := rowSelector(columns, pkColumn).
+		From(QuotedTableNameFromString(schema, table)).
+		Where(sq.Eq{quotedPK: pks}).
+		OrderBy(quotedPK).
+		ToSql()
+
+	if err != nil {
+		return nil, err
+	}
+
+	// This query must be a prepared query. If it is not, querying will use
+	// MySQL's plain text interface, which will scan all values into []uint8
+	// if we give it []interface{}.
+	stmt, err := db.Prepare(sql)
+	if err != nil {
+		return nil, err
+	}
+
+	defer stmt.Close()
+	rows, err := stmt.Query(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+func (c *CompressionVerifier) decompressAndHash(table string, columns []schema.TableColumn, rows *sql.Rows, tableCompression map[string]string) (map[uint64][]byte, error) {
+	defer rows.Close()
+	resultSet := make(map[uint64][]byte)
+	for rows.Next() {
+		rowData, err := ScanGenericRow(rows, 2)
+		if err != nil {
+			return nil, err
+		}
+
+		pk, err := rowData.GetUint64(0)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decompress the applicable columns and then hash them together
+		// to create a fingerprint. decodedRowData contains a map of all
+		// columns and associated decoded values by the index of the column
+		var decodedRowData map[int][]byte
+		for idx, column := range columns {
+			// Check if column is configured as compressed and decompress if necessary
+			if algorithm, ok := tableCompression[column.Name]; ok {
+				decodedColData, err := c.Decode(table, column.Name, algorithm, rowData[idx].([]byte))
+				if err != nil {
+					return nil, err
```

Comment:
How do you feel about ignoring errors here? Imagine we have corrupted data in a compressed column, i.e. data that can't be decompressed (for example because it was truncated). With your code here, we will never ever be able to copy that data. But (at this layer) we actually don't care about the data being decompressable, we only care about it being equal to the data in the other database. So as long as both the source and the target are the same (even if both are corrupted, as long as they are corrupted in the exact same way), we should not treat that as an error (if you agree, can you add a test for that?). I'd suggest to treat decompression errors as "this column wasn't meant to be decompressed anyway" and just store it in `decodedRowData` as raw compressed data (same as the `else` branch below) and then give it a second chance by comparing using the regular md5 approach.

Not sure if this actually ever happens in practice, but (if I remember correctly) our current Ruby implementation of this algorithm doesn't have that problem since it only decompresses if the fingerprints are not equal already.

Reply:
Great call out. If there is a decompression error then surely it will cause problems in other places so it should definitely be logged (and possible make some noise?). 

We can continue execution and yes just confirm equality, however, if we're only going down this path when the fingerprints of the SQL don't match (as you point out in your other comment), and now the data cannot be decompressed, then we won't be able to confirm the equality of the data, right? 🤔

Reply:
@hkdsun, how do you feel about this comment? i.e. ALWAYS decompressing makes it impossible to move data that is corrupted (while that was previously possible).

Reply:
@fw42 just confirmed with @hkdsun that moves are aborted if the data cannot be decompressed and the compressed hashes do not match. Planning to implement this here. What are your thoughts?

Reply:
Yes, that's the current behaviour in the Ruby implementation.